program/rust/src/rust_oracle.rs

```diff
@@ -38,41 +45,119 @@ use crate::c_oracle_header::{
 
+use crate::utils::pyth_assert;
+
 use super::c_entrypoint_wrapper;
+const PRICE_T_SIZE: usize = size_of::<pc_price_t>();
+const PRICE_ACCOUNT_SIZE: usize = size_of::<PriceAccountWrapper>();
+
 
 ///Calls the c oracle update_price, and updates the Time Machine if needed
 pub fn update_price(
     _program_id: &Pubkey,
-    _accounts: &[AccountInfo],
+    accounts: &[AccountInfo],
     _instruction_data: &[u8],
     input: *mut u8,
 ) -> OracleResult {
-    //For now, we did not change the behavior of this. this is just to show the proposed structure
-    // of the program
-    c_entrypoint_wrapper(input)
+    let [_funding_account_info, price_account_info, _sysvar_clock] = match accounts {
+        [x, y, z] => Ok([x, y, z]),
+        _ => Err(ProgramError::InvalidArgument),
+    }?;
+    //accounts checks happen in c_entrypoint
+    let account_len = price_account_info.try_data_len()?;
+    match account_len {
+        PRICE_T_SIZE => c_entrypoint_wrapper(input),
+        PRICE_ACCOUNT_SIZE => {
+            let c_ret_value = c_entrypoint_wrapper(input)?;
+            if c_ret_value == SUCCESSFULLY_UPDATED_AGGREGATE {
+                let mut price_account =
+                    load_account_as_mut::<PriceAccountWrapper>(price_account_info)?;
+                price_account.add_price_to_time_machine()?;
+            }
+            Ok(c_ret_value)
+        }
+        _ => Err(ProgramError::InvalidArgument),
+    }
 }
-/// has version number/ account type dependant logic to make sure the given account is compatible
-/// with the current version
-/// updates the version number for all accounts, and resizes price accounts
-pub fn update_version(
-    _program_id: &Pubkey,
-    _accounts: &[AccountInfo],
+fn send_lamports<'a>(
+    from: &AccountInfo<'a>,
+    to: &AccountInfo<'a>,
+    system_program: &AccountInfo<'a>,
+    amount: u64,
+) -> Result<(), ProgramError> {
+    let transfer_instruction = transfer(from.key, to.key, amount);
+    invoke(
+        &transfer_instruction,
+        &[from.clone(), to.clone(), system_program.clone()],
+    )?;
+    Ok(())
+}
+
+/// resizes a price accpunt so that it fits the Time Machine
+/// key[0] funding account       [signer writable]
+/// key[1] price account         [Signer writable]
+/// key[2] system program        [readable]
+pub fn resize_price_account(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
     _instruction_data: &[u8],
 ) -> OracleResult {
-    panic!("Need to merge fix to pythd in order to implement this");
-    // Ok(SUCCESS)
+    let [funding_account_info, price_account_info, system_program] = match accounts {
+        [x, y, z] => Ok([x, y, z]),
+        _ => Err(ProgramError::InvalidArgument),
+    }?;
+
+    check_valid_funding_account(funding_account_info)?;
+    check_valid_signable_account(program_id, price_account_info, size_of::<pc_price_t>())?;
+    pyth_assert(
+        check_id(system_program.key),
+        OracleError::InvalidSystemAccount.into(),
+    )?;
```

Comment:
I think you need to check that's a price account here : 
I recommended using `load_checked<pc_price_t>`

Reply:
There's a load_checked at the end, with a comment that explains why it is sufficient.
This is cleaner than having to accesss a smaller price account then a bigger one

Reply:
fwiw i think it's better to add a call to `load_checked` here in addition to your existing check. I agree your code as written is technically correct, but it's more prone to human error: future edits to this code could easily eliminate the necessary check.